man/man1/ccd2vel.1
.TH CCD2VEL 1NEMO "11 March 2022"

.SH "NAME"
ccd2vel \- convert map of wave/freq/velocities to doppler velocities

.SH "SYNOPSIS"
\fBccd2vel\fP [parameter=value]

.SH "DESCRIPTION"
\fBccd2vel\fP converts a map to a doppler velocity map. The input map
units can be in wavelength (\fBrestwave=\fP is needed), frequency (\fBrestfreq=\fP is needed)
or one of three doppler velocity frames of reference (radio, optical, relativistic).
The output map will be in one of those three doppler velocity frames of reference
(radio, optical, relativistic). Output units are normally km/s.
.PP
This program ignores references to being in a Barycenric, Heliocentric or LSR frame. See comments below in the
CASA and MIRIAD sections.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP
Optional input image file. No default.
.TP
\fBout=\fP
Output image file. No Default.
.TP
\fBfrom=rel|rad|opt\fP
Frame of reference (relativistic, radio, optical). When using \fBrestwave=\fP or \fBrestfreq=\fP this keyword is ignored.
[rel] 
.TP
\fBto=rel|rad|opt\fP
Frame of reference (relativistic, radio, optical) [rel] 
.TP
\fBrestwave=\fP
Rest wavelength, in the units of the map.  
Only one of restwave of restfreq can be used.
.TP
\fBrestfreq=\fP
Rest frequency, in the units of the map
.TP
\fBclip=\fP\fImin,max\fP
Clipping of output (none, or two values needed). Default: no clipping.
.TP
\fBunits=km/s\fP
Desired output units (km/s, m/s) [km/s]
.TP
\fBunits=\fP
If given, and no in/out is given, the three velocity conventions (radio, relativistic, optical)
are given in km/s. Default: not used.

.SH "EXAMPLES"
Here is how one converts a fits cube with an optical wavelengths 3rd axis
to a relatistic velocity field:
.nf
   fitsccd Ha.cube.fits  - | ccdmom - - mom=3 | ccd2vel - Ha-3.vel opt rel 6562.80 
.fi
and here an optical velocity field into a relativistic
.nf
   NGC2347.VoptFromHa.fits - | ccd2vel - NGC2347.VrelFromHa.ccd opt rel
.fi
and the different doppler conversion velocities for z=1:
.nf
    ccd2vel z=1
    vrad=149896
    vrel=179875
    vopt=299792
.fi

.SH "CASA"
From the \fBtclean\fP documentation on the \fBveltype\fP keyword:

Velocity type (radio, z, ratio, beta, gamma, optical)
For start and/or width specified in velocity, specifies the velocity definition

Options: 'radio','optical','z','beta','gamma'

NOTE: the viewer always defaults to displaying the 'radio' frame, 
but that can be changed in the position tracking pull down.

The different types (with F = f/f0, the frequency ratio), are:
.nf

                       Z = (-1 + 1/F)
                      RATIO = (F) *
                      RADIO = (1 - F)
                      OPTICAL == Z
                      BETA = ((1 - F2)/(1 + F2))
                      GAMMA = ((1 + F2)/2F) *
                      RELATIVISTIC == BETA (== v/c)
                      DEFAULT == RADIO
                      Note that the ones with an '*' have no real interpretation 
                      (although the calculation will proceed) if given as a velocity.

.fi

.SH "MIRIAD"
From the \fBVELSW\fP documentation:

VELSW is a Miriad task used to change the type of a spectral axis.
For example, a frequency axis could be changed to redshift or
relativistic velocity.

The resulting axis may be non-linear.  If so, it will be marked by a
code, e.g. VOPT-F2W, meaning that it is actually linear in frequency
(F) but is labelled as a spectral type that is linear in wavelength
(W).  Such non-linear axes are handled properly by the coordinate
routines.  However, when plotted, non-linear axes are only correct to
first-order at the reference point.

Switching between velocity and other spectral types, via the
Doppler formula, is only possible if a rest frequency is
recorded in the header.  It is also possible to change the
Doppler reference frame to barycentric or LSRK (only).

Note that velocity axes labelled as {VELO,FELO}-{HEL,LSR} in
Miriad have been Doppler-shifted to the specified frame
(barycentric or LSRK) as would be expected.  However, frequency
axes labelled as FREQ-{HEL,LSR} are actually topocentric, the
Doppler frame here refers only to the value of vobs recorded in
the header.  This convention also applies to FITS images written
and read by Miriad - the presence of vobs with FREQ-{HEL,LSR} in
the header of a Miriad image read from such a FITS file is
indicative.  However, there is a potential inconsistency when
FITS files generated by other packages are read into Miriad.
See option altspc below.

.nf
Keyword: axis
         The spectral type to change to:
           FREQ: frequency,
           ENER: photon energy (linear with frequency),
           WAVN: wavenumber (linear with frequency),
           VRAD: radio velocity (linear with frequency),
           WAVE: wavelength,
           VOPT: optical velocity (linear with wavelength),
           ZOPT: redshift (linear with wavelength),
           AWAV: air wavelength (applicable to optical regime only),
           VELO: apparent radial velocity (relativistic),
           BETA: relativistic beta factor (= VELO/c).
.fi

Note that there is no minimum-match on these four-letter
spectral types.  For backwards compatibility, 'frequency', 'radio',
and 'optical' (with min-match) are understood as
synonyms for 'FREQ', 'VRAD', and 'VOPT'.

AIPS-convention types, {FREQ,VELO,FELO}-{HEL,LSR}, are also
understood and will be translated.  Here VELO means radio
velocity (VRAD), and FELO is optical velocity (VOPT-F2W).
However, these types are deprecated, and there is no way to
switch to them explicitly.

An optional second value defines the Doppler frame to transform
to (min-match applies):
.nf
   BARYCENT: Solar System barycentre (aka BARYCENT{RE,ER,RIC}, or HEL),
   LSRK: Kinematic Local Standard of Rest.
.fi
The default for both the spectral type and the Doppler frame is
to leave them as is, performing only required translations.

.fi

.SH "SEE ALSO"

ccdsky(1NEMO), ccdmath(1NEMO), fitsccd(1NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
27-Apr-17	V0.3 Created	PJT
11-mar-11	V0.6 added doppler conversions for given z=	PJT
.fi


src/image/trans/ccd2vel.c
/* 
 * CCD2VEL: convert a map of frequencies or wavelengths to doppler velocity,
 *          in any frame (radio, optical, relativistic)
 *
 *      27-apr-2017     created        Peter Teuben
 *
 */


#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <image.h>

#include <mks.h>

string defv[] = {
  "in=\n            Optional input image file",
  "out=\n           Output image file",
  "from=rel\n       Input reference frame of reference, if converting from velocities (relativistic, radio, optical)",
  "to=rel\n         Output reference frame of reference (relativistic, radio, optical)",
  "restwave=\n      Input rest wavelength, in the units of the map, only if needed",
  "restfreq=\n      Input rest frequency, in the units of the map, only if needed",
  "clip=\n          Clipping of output (none, or two values needed)",
  "units=km/s\n     Desired output units (km/s, m/s)",
  "z=\n             Show conversions if no files used",
  "VERSION=0.6\n    11-mar-2022 PJT",
  NULL,
};

string usage = "convert map of wave/freq to doppler velocities";


static real c = c_MKS / 1000.0;    /* use km/s for now */

#define CVI(x,y,z)  CubeValue(iptr,x,y,z)


#define ONE2ONE  0x00

#define WAV2REL  0x01
real wav2rel(real w,real w0)
{
  return c * (w*w-w0*w0) / (w*w+w0*w0);
}

#define FRQ2REL  0x02
real frq2rel(real f,real f0)
{
  return c * (f0*f0-f*f) / (f0*f0+f*f);
}

#define OPT2RAD  0x04
real opt2rad(real v)
{
  return v / (1 + v/c);
}

#define OPT2REL  0x08
real opt2rel(real v)
{
  real q = 1 + v/c;
  return c * (q*q-1)/(q*q+1);
}

#define RAD2OPT  0x10
real rad2opt(real v)
{
  return v / (1 - v/c);
}

#define RAD2REL  0x20
real rad2rel(real v)
{
  real q = 1 - v/c;
  return c * (1-q*q)/(1+q*q);
}

#define REL2OPT  0x40
real rel2opt(real v)
{
  real z = v/c;
  return c*(sqrt((1+z)/(1-z)) - 1);
}

#define REL2RAD 0x80
real rel2rad(real v)
{
  real z = v/c;
  return c*(1 - sqrt((1-z)/(1+z)));
}

#define WAV2OPT  0x0100
real wav2opt(real w,real w0)
{
  return c * (w/w0 - 1.0);
}

#define FRQ2RAD  0x0200
real frq2rad(real f,real f0)
{
  return c * (1.0 - f/f0);
}


// optical to z
real opt2z(real vopt)
{
  return vopt/c;
}

// z to optical
real z2opt(real z)
{ 
  return c*z;
}

// z to radio
real z2rad(real z)
{
  return opt2rad(z2opt(z));
}




void nemo_main()
{
    stream  instr, outstr;
    int     ix, iy, iz, nx, ny;
    int     mode1, mode2;
    imageptr iptr=NULL;
    string to    = getparam("to");
    string from  = getparam("from");
    bool Qclip, Qfirst, Qlinear = FALSE;
    real isca = 1.0, ioff = 0.0;
    real osca = 1.0, ooff = 0.0;
    real zsca = 1.0, zoff = 0.0;
    real vel, vmin, vmax, clip[2];
    real restwave = -1;
    real restfreq = -1;

    if (hasvalue("restwave")) restwave = getrparam("restwave");
    if (hasvalue("restfreq")) restfreq = getrparam("restfreq");
    Qclip = hasvalue("clip");
    if (Qclip) {
      if (nemoinpr(getparam("clip"),clip,2) != 2) error("Two values needed for clip=");
      dprintf(0,"Clipping on %g %g\n",clip[0], clip[1]);
    }

    if (restwave > 0)
      mode1 = WAV2REL;
    else if (restfreq > 0)
      mode1 = FRQ2REL;
    else if (streq(from,"opt"))
      mode1 = OPT2REL;
    else if (streq(from,"rad"))
      mode1 = RAD2REL;
    else if (streq(from,"rel"))
      mode1 = ONE2ONE;
    else
      error("Unknown from= frame");

    if (streq(to,"rad"))
      mode2 = REL2RAD;
    else if (streq(to,"opt"))
      mode2 = REL2OPT;
    else if (streq(to,"rel"))
      mode2 = ONE2ONE;
    else
      error("Unknown to= frame");

    /* set the input and output modes */
    dprintf(0,"MODES: 0x%x  and 0x%x\n",mode1,mode2);
    /* then see if there are shortcuts */
    if (mode1 == OPT2REL && mode2 == REL2OPT) { mode1 = mode2 = ONE2ONE;}
    if (mode1 == RAD2REL && mode2 == REL2RAD) { mode1 = mode2 = ONE2ONE;}
    if (mode1 == RAD2REL && mode2 == REL2OPT) { mode1 = RAD2OPT; mode2 = ONE2ONE;}
    if (mode1 == OPT2REL && mode2 == REL2RAD) { mode1 = OPT2RAD; mode2 = ONE2ONE;}
    if (mode1 == WAV2REL && mode2 == REL2OPT) { mode1 = WAV2OPT; mode2 = ONE2ONE;}
    if (mode1 == FRQ2REL && mode2 == REL2RAD) { mode1 = FRQ2RAD; mode2 = ONE2ONE;}        
    dprintf(0,"MODES: 0x%x  and 0x%x\n",mode1,mode2);
    dprintf(0,"Using c=%f km/s\n",c);
    
    if (hasvalue("in") && hasvalue("out")) {    
      instr = stropen(getparam("in"), "r");
      read_image( instr, &iptr);
      nx = Nx(iptr);
      ny = Ny(iptr);
      Qfirst = TRUE;
      for (iy=0; iy<ny; iy++)
	for (ix=0; ix<nx; ix++) {
	  vel = MapValue(iptr,ix,iy);
	  if (Qlinear)
	    vel = zsca * vel + zoff;	  /* linear velocity conversion (radio,optical) */
	  else {
	    if (mode1 == WAV2REL)
	      vel = wav2rel(vel, restwave);
	    else if (mode1 == WAV2OPT)
	      vel = wav2opt(vel, restwave);
	    else if (mode1 == FRQ2REL)
	      vel = frq2rel(vel, restfreq);
	    else if (mode1 == FRQ2RAD)
	      vel = frq2rad(vel, restfreq);
	    else if (mode1 == OPT2REL)
	      vel = opt2rel(vel);
	    else if (mode1 == RAD2REL)
	      vel = rad2rel(vel);
	    else if (mode1 == OPT2RAD)
	      vel = opt2rad(vel);
	    else if (mode1 == RAD2OPT)
	      vel = rad2opt(vel);
 
	    if (mode2 == REL2OPT)
	      vel = rel2opt(vel);
	    else if (mode2 == REL2RAD)
	      vel = rel2rad(vel);
	  }

	  if (Qclip)
	    if (vel < clip[0] || vel > clip[1]) vel = 0.0;

	  if (Qfirst) {
	    if (vel != 0.0) {
	      vmin = vmax = vel;
	      Qfirst = FALSE;
	    }
	  }
	  if (vel != 0.0) {
	    vmin = MIN(vel,vmin);
	    vmax = MAX(vel,vmax);
	  }
	  MapValue(iptr,ix,iy) = vel;
	}
      MapMin(iptr) = vmin;
      MapMax(iptr) = vmax;
      dprintf(0,"MinMax: %g %g\n",vmin,vmax);
      outstr = stropen(getparam("out"), "w");
      write_image(outstr, iptr);
    } else if (hasvalue("z")) {
      real z = getrparam("z");
      real vopt = z2opt(z);
      real vrad = z2rad(z);
      printf("z=%g\n",z);
      printf("vrad=%g\n",vrad);
      printf("vrel=%g\n",opt2rel(vopt));
      printf("vopt=%g\n",vopt);
    }

    
}
